libraries/AP_NavEKF/AP_NavEKF_Source.h

```diff
@@ -0,0 +1,107 @@
+#pragma once
+
+#include <AP_Param/AP_Param.h>
+
+#define AP_NAKEKF_SOURCE_COUNT 3    // three banks of sources
+
+class AP_NavEKF_Source
+{
+
+public:
+    // Constructor
+    AP_NavEKF_Source();
+
+    /* Do not allow copies */
+    AP_NavEKF_Source(const AP_NavEKF_Source &other) = delete;
+    AP_NavEKF_Source &operator=(const AP_NavEKF_Source&) = delete;
+
+    enum class SourceXY {
+        NONE = 0,
+        // BARO = 1 (not applicable)
+        // RANGEFINDER = 2 (not applicable)
+        GPS = 3,
+        BEACON = 4,
+        OPTFLOW = 5,
+        EXTNAV = 6,
+        WHEEL_ENCODER = 7
+    };
+
+    enum class SourceZ {
+        NONE = 0,
+        BARO = 1,
+        RANGEFINDER = 2,
+        GPS = 3,
+        BEACON = 4,
+        // OPTFLOW = 5 (not applicable)
+        EXTNAV = 6
+    };
+
+    enum class SourceYaw {
+        NONE = 0,
+        COMPASS = 1,
+        EXTERNAL = 2,
+        EXTERNAL_COMPASS_FALLBACK = 3
+    };
+
+    // initialisation
+    void init();
+
+    // get current position source
+    SourceXY getPosXYSource() const { return _active_source.initialised ? _active_source.posxy : (SourceXY)_source[0].posxy.get(); }
+    SourceZ getPosZSource() const { return _active_source.initialised ? _active_source.posz : (SourceZ)_source[0].posz.get() ; }
+
+    // set position and velocity sources to either 0=primary, 1=secondary, 2=tertiary
+    void setPosVelXYZSource(uint8_t source_idx);
+
+    // get/set velocity source
+    SourceXY getVelXYSource() const { return _active_source.initialised ? _active_source.velxy : (SourceXY)_source[0].velxy.get(); }
```

Comment:
@peterbarker this would force us to make getVelXYSource a non-const function.  If people really don't like this I can change it but I like having the accessors as const.

Reply:
I've removed the initialised check so it simply returns the active source.  It's possible that this could return NONE in case init() has not yet been called but I think this should be OK.